src/java/org/apache/cassandra/index/sai/utils/PrimaryKeyWithSortKey.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.utils;
+
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+/**
+ * A PrimaryKey with one piece of metadata. The metadata is not used to determine equality or hash code, but it is used
+ * to compare the PrimaryKey objects.
+ * Note: this class has a natural ordering that is inconsistent with equals.
```

Comment:
This natural ordering inconsistent with equals is worrying me a bit. Can't we make it a separate `Comparator`? This class could just define an abstrtact factory for such comparator.

Reply:
This is a copy of the relevant part of the JavaDoc for `Comparator`:

```
It is generally the case, but not strictly required that (compare(x, y)==0) == (x.equals(y)). Generally speaking, any comparator that violates this condition should clearly indicate this fact. The recommended language is "Note: this comparator imposes orderings that are inconsistent with equals."
```

I share your concern. I tried to make it work where the PrimaryKey was just a field on this class, but that change will require refactoring the `Plan.Executor` to return something other than a PrimaryKey:

```java
    public interface Executor
    {
        Iterator<? extends PrimaryKey> getKeysFromIndex(Expression predicate);
        Iterator<? extends PrimaryKey> getTopKRows();
        Iterator<? extends PrimaryKey> getTopKRows(RangeIterator keys);
    }
```

I looked into this as a solution, but it was going to require a larger refactoring. If we want to refactor Plan, we can also refactor the `PrimaryKeyWithSortKey`.